utils/docs/gsymfix.1.in
.TH gsymfix 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
gsymfix \- automatically fix common issues with gEDA symbols

.SH SYNOPSIS
.B gsymfix [
.I options
.B ]
.I symbol_file
.SH DESCRIPTION
.B gsymfix
reads a gEDA symbol file and does the following:
.IP \(bu
prints a warning if there is no pinnumber attribute on a pin;
.IP \(bu
Renumbers pinseq attributes on all pins, starting with 1;
.IP \(bu
Adds a pinseq attribute to a pin if none exists;
.IP \(bu
Adds a pinlabel=n_a attribute to a pin if none exists;
.IP \(bu
Adds a pintype=io attribute to a pin if none exists;
.IP \(bu
Adds a numslots=0 attribute to the symbol if none exists;
.IP \(bu
Adds a device=XXX attribute to the symbol if none exists;
.IP \(bu
Adds a footprint=XXX attribute to the symbol if none exists.
.PP
The idea here is to fix common cut-and-paste issues and fix up
the symbol so that it will pass
.BR gsymcheck (1)
with no errors or
warnings.

.SH OPTIONS
.B gsymfix
accepts the following options:
.TP
.BR -h ", " \-\-help
Displays this help message.
.TP
.B \-\-verbose
Enables verbose output.
.TP
.B \-\-vverbose
Enables *very* verbose output.
.TP
.B \-\-version
Shows the version of this program.

.SH AUTHOR
Mike Skerritt <mike@acornpacket.com>.

.SH SEE ALSO
.BR gschem (1),
.BR gsymcheck (1)

.SH COPYRIGHT
.nf
Copyright \(co 1999-2014 gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2014 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 2.0.8 - Rampage Edition

* Gschem has improved performances, additional toolbars, improved stability,
  enhanced appearance and improved memory management. The main menu has been
  revised to improve clarity and organization. New features have been added
  such a Invert Selection, Zoom-To Selection,

* Libgeda extensively revised to provide a more generic API. The library
  has improved performance from previous versions and uses less memory,
  The was mostly achieved by eliminating the needless passing of parameters.
  All Object types are now object oriented to improve memory management
  by off-loading responsibility to glib. The API has been extended to
  provide more functionality for applications. The configuration storage
  systems provided by Libgeda have been revised, mostly the new system,
  which no longer utilizes gnome based virtual file systems, aka gvfs. The
  use of GFile exponentially increased dependencies and was eliminated.

* Symbol libraries significantly expanded.

* A new library, Libgedathon provides an API for Python scripts.

* A new library, Libgedawidget provides an "Gtk" like widgets to enhance
  geda-gaf applications and extend support of code obsoleted by Gtk
  developers.

* The build system has been "heavily" modified and the source code
  completely reorganized. New macros provide better organization and
  greater flexibility.

* New install scripts, geda-install and geda-install-bin facilitate
  installation from local builds. (Note that geda-install-bin replaces
  ginstall_local)

Notable changes in gEDA/gaf 2.0.7.8 - Rampage Edition

* The Coordinates Dialog allows direct entering of World Coordinates.
  If an XY coordinate is "entered" when not in an Action, the cursor
  will be positioned over the point so that "hot-keys" can be used,
  to begin an action starting from an exact point. Parentheses for
  the coordinates is optional.

* Gschem now supports bidirectional Drag & Drop. Contents of Schematic
  drawings can be "dragged" from one gschem window to any other instance
  of gschem. The target can also be the World Entry of the Coordinates
  Dialog of the same or another instances of gschem. Dropping schematic
  objects on the World Entry requires that a value be entered immediately
  afterward for the dropped object to be inserted into the drawing at the
  specified point. Dropping a ".sch" file results in the file being loaded
  into new window. The results of dropping a ".sym" file depends on where
  the file is located. If the symbol file is in a Library source directory
  then the symbol is inserted into the current Schematic, otherwise the
  symbol file is opened in a new window for editing as a Symbol definition.

* "Save As" can now optionally add the ".sch" or ".sym" file extension.
  based on the filter selection.

* Generated color images file can now be inverted. User can choose between
  the Display or the Print Color map on Write Image Dialog.

* Gschem is more tolerent of errors in RC files.

* Message dialogs are more consistence.

Notable changes in gEDA/gaf 2.0.7 - Rampage Edition
=================================

* The Gdk cursor to be use over the Drawing Area is now user
  configurable via the Preference dialog.

* New Path drawing tool from Master bramch implemented.

* Many user settings are no long dependent on RC entries and are
  saved at the end of each session if the "save settings on exit"
  option is enabled, allowing users to change settings on-the-fly
  without permenently changing their configuration.

* New "Deselect" mode can be used to remove objects from a selection
  set.

* The color of grips, text markers, and cues can be changed using a
  color selector dialog. In addition the size of grips, markers and
  cues can be customized by the user in the Preferences dialog.

* Gschem no longer closes the GUI when the last file is closed.

* Gschem's "Save As" can now optionaly add the ".sch" file extension.

* Refdes number are reset when using buffers and the clipboard.


Notable changes in gEDA/gaf 2.0.6 - Rampage Edition
=================================

* GUI Enhancments for gschem GUI with better consistency between
  dialogs. Most dialog are now modeless and react to selection
  changes. 

* Component selection enhanced to facilitate selection of library
  symbols with larger libraries. The selection dialog provides a
  more advanced interface to allow catagorization of symbols.

* Symbol libraries significantly expanded. A new "style" system is
  introduced. The new style scheme provides a "standardization" of
  symbols names. Filtering based on style is supported with user
  configurable definitions and GUI options.

* Additional support scripts have been added to supplement the suite.
  Scripts are provided to batch test and repair libraries.

* Man pages have now been added for all geda-gaf programs.

Notable changes in gEDA/gaf 2.0.1 - Rampage Edition
=================================

* GUI's for gschem and gattrib have received Major Make-Overs with
  toolbars and additional menu options, check box menu items, and
  more. All user options can now be set in Gschem via a multi-tab
  dialog with custom controls without adding GTK+3 as dependent or
  any other new dependencies. Dialogs uniformity has been improved. 

* Functionality of gattrib has been greatly extended including a
  new Search and Replace function complete with session search
  history!

* Gattrib no long launches into a a file selector dialog, files can
  now be loaded or closed via menu or toolbars, The editing bar is
  now fully functional and mouse menus added. Embedded GTK-Sheet has

utils/scripts/gsymfix.pl
#!/usr/bin/perl -w

my $version = "0.8";

# This script reads a geda symbol and does the following:
# (...see the usage routine below...)

# for parsing input options
use Getopt::Long;

# don't allow -he to be interpreted as --help
$Getopt::Long::autoabbrev=0;

&GetOptions(("help|h" => \&usage,
	     "verbose" => \$verbose,
	     "vverbose" => \$vverbose,
	     "version" => \&version
	     ));

usage() if $Getopt::Long::error;
usage() unless @ARGV;

my $found_pinnumber_attr = 0;
my $found_pinseq_attr    = 0;
my $found_pintype_attr   = 0;
my $found_pinlabel_attr  = 0;
my $found_numslots_attr  = 0;
my $found_device_attr    = 0;
my $found_footprint_attr = 0;
my $seqcnt               = 0;
my $skip_line_out        = 0;
my $file_line            = 0;

my $st_scan              = 0;
my $st_pin_start         = 1;
my $st_pin_body          = 2;

my $state                = $st_scan;	# intial state machine state

if($vverbose) { $verbose = 1; }

# make sure the input schematic exists and we can read it
my $fname=shift(@ARGV);
print "\nProcessing symbol file $fname\n";
open(NETLIST,"$fname") or die "Can't open $fname: $!\n";

# open output netlist
$outfname="$fname.fix";
open(OUTNET,">$outfname") or die "Can't open $outfname: $!\n";

# parse netlist
while($line = <NETLIST>) {
  $file_line++;
  #==========================
  if( $state == $st_scan ) {
    if( $line =~ /^P/) { 
      $state = $st_pin_start; 
      if($vverbose){print "Pin start...\n";}
    }
    elsif( $line =~ /^numslots=/) { 
      $found_numslots_attr = 1;
    }
    elsif( $line =~ /^device=/) { 
      $found_device_attr = 1;
    }
    elsif( $line =~ /^footprint=/) { 
      $found_footprint_attr = 1;
    }
    print OUTNET $line;
  }
  #==========================
  elsif( $state == $st_pin_start ) {
    if( $line =~ /^{/) {
      $state = $st_pin_body;
      $found_pinnumber_attr = 0;
      $found_pinseq_attr    = 0;
      $found_pintype_attr   = 0;
      $found_pinlabel_attr  = 0;
      print OUTNET $line;
    } 
    else {
      print STDERR "*** ERROR: No pin left bracket found at line $file_line\n";
      exit(-1);
    }
  }
  #==========================
  elsif( $state == $st_pin_body ) {
    #----------------------
    if( $line =~ /^pinnumber/) {
      $found_pinnumber_attr = 1;
      $pin_num = $line;
      $pin_num =~ s/\n//;
      $pin_num =~ s/^pinnumber=//;
      $pinnumber_attr_line = $prev_line;
      if($vverbose){print "  pinnumber attribute found ($pin_num)\n";}
    }
    #----------------------
    elsif( $line =~ /^pinseq/) {
      $found_pinseq_attr = 1;
      if($vverbose){print "  Pinseq attribute found\n";}
      $seqcnt++;
      if($vverbose){print "  Renumbering attr pinseq=$seqcnt\n";}
      print OUTNET "pinseq=$seqcnt\n";
      $skip_line_out++;
    }
    #----------------------
    elsif( $line =~ /^pintype/) {
      $found_pintype_attr = 1;
      if($vverbose){print "  pintype attribute found\n";}
    }
    #----------------------
    elsif( $line =~ /^pinlabel/) {
      $found_pinlabel_attr = 1;
      if($vverbose){print "  pinlabel attribute found\n";}
    }
    #----------------------
    elsif( $line =~ /^}/) {
      $state = $st_scan;
      if( $found_pinnumber_attr == 0 ) {
        print "*** WARNING: no pinum attribute found at line $file_line\n";
      }
      if( $found_pinseq_attr == 0 ) {
	$seqcnt++;
	if($verbose){print "  Pin $pin_num: Adding attr pinseq=$seqcnt\n";}
	print OUTNET $pinnumber_attr_line;
	print OUTNET "pinseq=$seqcnt\n";
      }
      if( $found_pintype_attr == 0 ) {
	if($verbose){print "  Pin $pin_num: Adding attr pintype=io\n";}
	print OUTNET $pinnumber_attr_line;
	print OUTNET "pintype=io\n";
      }
      if( $found_pinlabel_attr == 0 ) {
	if($verbose){print "  Pin $pin_num: Adding attr pinlabel=n_a\n";}
	print OUTNET $pinnumber_attr_line;
	print OUTNET "pinlabel=n_a\n";
      }
      if($vverbose){print "Pin end...\n";}
    }
    if( $skip_line_out ) {
      $skip_line_out = 0;
    } else {
      print OUTNET $line;
    }
    $prev_line = $line;		# Save line for next pass
  }
}

if( $found_numslots_attr == 0 ) {
  if($verbose){print "  Adding attr numslots=0\n";}
  print OUTNET "T 600 100 9 10 0 0 0 0 1\n";
  print OUTNET "numslots=0\n";
}
if( $found_device_attr == 0 ) {
  if($verbose){print "  Adding attr device=none\n";}
  print OUTNET "T 600 100 9 10 0 0 0 0 1\n";
  print OUTNET "device=none\n";
}
if( $found_footprint_attr == 0 ) {
  if($verbose){print "  Adding attr footprint=unknown\n";}
  print OUTNET "T 600 100 9 10 0 0 0 0 1\n";
  print OUTNET "footprint=unknown\n";
}

close(NETLIST);
close(OUTNET);

print STDERR "Output written to file $outfname\n\n";
exit;


#######################################################################
#
# Subroutines
#
#######################################################################

#---------------------------------
# usage()
#
# prints program usage
#---------------------------------

sub usage {
  my $pname = $0;
  $pname =~ s/.*\///g;

  print "\nUsage:\n\n";
  print "    $pname [option] symbol_file\n";
  print "\n";
  print "    $pname reads a geda symbol file and does the following:\n";
  print "\t- Prints a warning if there is no pinnumber attribute on a pin\n";
  print "\t- Renumbers pinseq attributes on all pins, starting with 1\n";
  print "\t- Adds a pinseq attribute to a pin if none exists\n";
  print "\t- Adds a pinlabel=n_a attribute to a pin if none exists\n";
  print "\t- Adds a pintype=io attribute to a pin if none exists\n";
  print "\t- Adds a numslots=0 attribute to the symbol if none exists\n";
  print "\t- Adds a device=XXX attribute to the symbol if none exists\n";
  print "\t- Adds a footprint=XXX attribute to the symbol if none exists\n";
  print "    The idea here is to fix common cut and paste issues and fix up\n";
  print "    the symbol so that it will pass gsymcheck with no errors or\n";
  print "    warnings.\n";
  print "\n";
  print "    $pname accepts the following options:\n";
  print "\n";
  print "    --help, -h Displays this help message.\n";
  print "\n";
  print "    --verbose   Enables verbose output.\n";
  print "\n";
  print "    --vverbose  Enables *very* verbose output.\n";
  print "\n";
  print "    --version   Shows the version of this program.\n";
  print "\n";
  print "    $pname was written by Mike Skerritt <mike\@acornpacket.com>\n";
  print "\n\n";
  exit;
}

#---------------------------------
# version()
#
# prints program version
#---------------------------------

sub version {
  my $pname = $0;
  $pname =~ s/.*\///g;
  print "$pname ($0):  Version $version\n";
}


